### PR TITLE
smile :)

### PR DESCRIPTION
**Pull request by cole-h**
Today is a good day.

### CODE REVIEW

**Comment by timokau** [comment]
Indeed it is :upside_down_face: 

/marvin opt-in

**Comment by timokau** [comment]
Oh, you would have to do that. Well, let me cheat.

**Comment by cole-h** [comment]
/marvin opt-in

EDIT: So close!

**Comment by timokau** [comment]
Hah, I probably outran it by a millisecond or two.

**Comment by timokau** [comment]
I keep mixing that up

/state needs_review

**Comment by timokau** [comment]
/status needs_review
... you'd think I know the commands

**Comment by timokau** [comment]
Oh, it crashed. Good thing I tried it here first.

**Comment by timokau** [comment]
Third time's the charm.
/status needs_review

File: smile
Comment (timokau) [inline smile]: Needs more emoji.

**Review by timokau — CHANGES_REQUESTED** [review]

**Comment by timokau** [comment]
Actually ascii art might be fine.
/status needs_review

**Review by timokau — CHANGES_REQUESTED** [review]
Or is it?

**Review by timokau — COMMENTED** [review]
I think I like it.
/status needs_merge

**Comment by cole-h** [comment]
I've added an ASCII art smiley. PTAL.

```
                    *****************
               ******               ******
           ****                           ****
        ****                                 ***
      ***                                       ***
     **           ***               ***           **
   **           *******           *******          ***
  **            *******           *******            **
 **             *******           *******             **
 **               ***               ***               **
**                                                     **
**       *                                     *       **
**      **                                     **      **
 **   ****                                     ****   **
 **      **                                   **      **
  **       ***                             ***       **
   ***       ****                       ****       ***
     **         ******             ******         **
      ***            ***************            ***
        ****                                 ****
           ****                           ****
               ******               ******
                    *****************

(taken from https://www.asciiart.eu/computers/smileys)

```

**Comment by timokau** [comment]
:laughing: 
/status needs_review
That makes me think that pushes should reset `needs_merge` -> `needs_review`.

**Comment by timokau** [comment]
It should do that now. Love the ascii art!
/status needs_merge

**Comment by cole-h** [comment]
Looks like it works!

File: frown
Comment (timokau) [inline frown]: :(
At least the proportions are appropriate :)

**Comment by timokau** [comment]
Well, that should be enough bot development for today. Thanks for the help :)

**Comment by cole-h** [comment]
Happy to! If you need more ~~victims~~ testers, feel free to ping. I'm on IRC most of the time ;-)

**Comment by timokau** [comment]
Thanks! I think at this point its probably best to wait for some real-world usage and see how well it works for people.

File: frown
Comment (timokau) [inline frown:14]:
```
  , ; ,   .-'"""'-.   , ; ,
  \\|/  .'         '.  \|//
   \-;-/   ()   ()   \-;-/
   // ;               ; \\
  //__; :.         .; ;__\\
 `-----\'.'-.....-'.'/-----'
        '.'.-.-,_.'.'
jgs       '(  (..-'
            '-'
``` (same [source](https://www.asciiart.eu/computers/smileys))

Testing a fix for https://github.com/nix-community/infra/pull/30#issuecomment-645673503.

**Review by timokau — CHANGES_REQUESTED** [review]

**Comment by timokau** [comment]
Oh, I forgot again.

**Review by timokau — APPROVED** [review]

File: frown
Comment (timokau) [reply frown:14]: \o/

**Review by timokau — CHANGES_REQUESTED** [review]

**Comment by timokau** [comment]
Good bot.

**Comment by timokau** [comment]
(trying to get the bot to request a review)
/status needs_merge

**Comment by cole-h** [comment]
Assuming you're trying to get it to request you as the reviewer: Is there a chance that the bot doesn't support *re*-requesting, since you've already "reviewed" this? (Also note that authors cannot "review" their own PR if you were trying to request me...)

**Comment by timokau** [comment]
Yes I'm trying to request me. Its possible that it may be an issue that I have an outstanding review, that's one of the things I want to test. But for now the issues are more mundane (i.e. related to getting the right API request and juggling githubs data).
/status needs_merge

**Comment by timokau** [comment]
I'm surprised that `mypy` can't detect wrong argument order of `hasattr`...
/status needs_merge

**Comment by timokau** [comment]
Hopeful for this one
/status needs_merge

**Comment by timokau** [comment]
I'm sure you can do it marvin!
/status needs_merge

**Comment by timokau** [comment]
Well, this time though.
/status needs_merge

**Comment by timokau** [comment]
Very close, just didn't bring your credentials. It'll work next time.
/status needs_merge

**Comment by timokau** [comment]
What if we try to do it again?
/status needs_merge

Edit: Nothing seems reasonable.

**Comment by cole-h** [comment]
If you don't already, I think removing some flags (like `needs_merge`) on conversion to and from a draft PR is a good idea.

**Comment by timokau** [comment]
Yeah, agreed. We're review-bound, so I'm tending towards automating going "down the chain" as much as possible while keeping going "up the chain" explicit.

**Comment by timokau** [comment]
I'll focus on the review requests first though, since that will provide an actual incentive to use it ;)

**Comment by timokau** [comment]
@cole-h I temporarily [signed you up](https://github.com/timokau/marvin-mk2/pull/21/files#diff-7cc9cf4bc2576f1fe8e54fd6f45d2ef5R49) for review requests to test the selection ;) Hopefully it'll request you in 50% of the cases now. Doesn't make sense to request you for your own PR of course, but that's now the point of the test :smile: 
/status needs_merge

**Comment by timokau** [comment]
I should calculate how often I have to run this until I can be 99.99% sure there is a bug
/status needs_merge

**Comment by timokau** [comment]
0.5 ** 5 = 0.03125
Getting unlikely
/status needs_merge

**Comment by timokau** [comment]
```
2020-06-19T20:05:41.025436+00:00 app[web.1]: Shuffled: ['cole-h', 'timokau']
2020-06-19T20:05:41.025549+00:00 app[web.1]: Shuffled: ['cole-h', 'timokau']
2020-06-19T20:05:41.025641+00:00 app[web.1]: Shuffled: ['cole-h', 'timokau']
2020-06-19T20:05:41.025720+00:00 app[web.1]: Shuffled: ['cole-h', 'timokau']
2020-06-19T20:05:41.025798+00:00 app[web.1]: Shuffled: ['cole-h', 'timokau']
2020-06-19T20:05:41.025874+00:00 app[web.1]: Shuffled: ['cole-h', 'timokau']
2020-06-19T20:05:41.025950+00:00 app[web.1]: Shuffled: ['cole-h', 'timokau']
2020-06-19T20:05:41.026028+00:00 app[web.1]: Shuffled: ['cole-h', 'timokau']
2020-06-19T20:05:41.026105+00:00 app[web.1]: Shuffled: ['cole-h', 'timokau']
2020-06-19T20:05:41.026181+00:00 app[web.1]: Shuffled: ['timokau', 'cole-h']
```

Python's `random`, are you okay?

**Comment by cole-h** [comment]
https://xkcd.com/221/

**Comment by timokau** [comment]
Okay seems like that actually works okay and I just forgot about what you said [previously](https://github.com/timokau/marvins-playground/pull/21#issuecomment-646817157) :facepalm: 

It told me it would've picked you though, so :shrug:

**Comment by timokau** [comment]
Should no longer attempt to request reviews from the PR author.
/status needs_merge

**Comment by timokau** [comment]
And it doesn't :)

**Review by timokau — CHANGES_REQUESTED** [review]
For some reason the changes requested -> needs_work didn't work on https://github.com/NixOS/nixpkgs/pull/89235

**Comment by timokau** [comment]
https://github.com/timokau/marvin-mk2/pull/25
Too bad, so commands in reviews won't work after all in this deploy.

**Comment by timokau** [comment]
Doing reviewery things.

**Review by timokau — COMMENTED** [review]
Some feedback

**Review by timokau — CHANGES_REQUESTED** [review]
Some other feedback (not sure if "Comment" reviews should trigger the status change).

**Comment by timokau** [comment]
@cole-h, please leave a comment. Testing https://github.com/timokau/marvin-mk2/issues/29 (I'm finishing up for today though, so I won't be responsive for now).

**Review by cole-h — COMMENTED** [review]
This kind of comment?

**Comment by cole-h** [comment]
Or this kind of comment?

**Comment by cole-h** [comment]
I'm not really a fan of a comment implicitly changing state (e.g. without the presence of an explicit `/marvin do_something`). As you can see above, I contributed nothing of value that would require another review. I feel like it adds noise where there doesn't need to be any.

**Comment by timokau** [comment]
Yes, its a heuristic. But the difference between `awaiting_reviewer` and `awaiting_changes` is only important when the PR is going stale. `awaiting_changes` is basically  the sink for abandoned PRs we can forget about. Stale `awaiting_reviewer` PRs will be triaged and assigned a new reviewer.

Since I want to avoid letting PRs slip through the cracks, I'd rather put them into the review queue one time too much. A reviewer can then set it back to `awaiting_changes`, where it'll stay as long as the PR author is inactive. The issue with explicit commands is that
- people forget to use them or they
- don't use them because they're perceived as "rude" (when sending a PR "down the ladder") and
- they add additional noise that actually reaches you through notifications.

So in summary, people really only have to care about the `needs_reviewer` and `needs_merger` state. They can mentally combine `awaiting_*` to `in_review`. The differences are only important for automatic triage, where some false-positives are okay. The goals are to avoid PRs getting stuck and optimize for the "reviewer with commit bit" bottleneck.

Given this justification, do you still think explicit commands would be better?

**Comment by timokau** [comment]
In the new model, explicit commands are only needed for opt-in, non-committer review (`-> needs_merger` transitions are still explicit to optimize for the bottleneck) and when the heuristic fails (for example a PR author wants to request another reviewer with `/status needs_reviewer`).

**Comment by timokau** [comment]
(Something seems to be wrong here, not sure what the bot is doing. Good thing we're having this conversation here).

**Comment by cole-h** [comment]
I think the justification is fair, now that you've explained it. Not to say I like it (heuristics give me the heebie-jeebies), but it's fair.

**Comment by timokau** [comment]
> I think the justification is fair, now that you've explained it. Not to say I like it (heuristics give me the heebie-jeebies), but it's fair.

Good :) Lets see how it works in practice. An alternative would be to always make the `needs_changes` transitions explicit, but that would need a bit more reviewer capacity. `needs_changes` would then kind of be a soft stale label.

> (Something seems to be wrong here, not sure what the bot is doing. Good thing we're having this conversation here).

Ohh I actually had a command in my message without realizing it. Maybe it was better to require them to be in their own line, not sure. Either way, no bug here.

**Comment by timokau** [comment]
Good bot! (Well, `awaiting_changes` was not intended but assigning reviewers is nice).

**Comment by timokau** [comment]
Testing some changes in determining reviewer availability (https://github.com/timokau/marvin-mk2/issues/71).
/status needs_reviewer

**Comment by timokau** [comment]
That worked surprisingly well on first try. Good bot.
/status awaiting_changes

**Comment by timokau** [comment]
Didn't work quote that well after all: https://github.com/timokau/marvin-mk2/pull/80

**Pull request closed without merging**
